src/resource_clients/request_queue.ts

```diff
@@ -96,6 +97,102 @@ export class RequestQueueClient extends ResourceClient {
+        let unprocessedRequests: Array<UnprocessedRequest> = [];
+        for (let i = 0; i < 1 + this.httpClient.maxRetries; i++) {
+            const response = await this.httpClient.call({
+                url: this._url('requests/batch'),
+                method: 'POST',
+                data: remainingRequests,
+                params: this._params({
+                    forefront: options.forefront,
+                    clientKey: this.clientKey,
+                }),
+            });
+            const data = response.data.data as RequestQueueClientBatchAddRequestsResult;
+            processedRequests = [...processedRequests, ...data.processedRequests];
+            unprocessedRequests = data.unprocessedRequests;
+
+            // Get unique keys of all requests processed so far
+            const processedRequestsUniqueKeys = processedRequests.map(({ uniqueKey }) => uniqueKey);
+            // Requests remaining to be processed are the all that remain
+            // @ts-expect-error Undefined id - it seems to be added by ow validation
+            remainingRequests = requests.filter(({ uniqueKey }) => !processedRequestsUniqueKeys.includes(uniqueKey));
+
+            // Stop if all requests have been processed
+            if (remainingRequests.length === 0) {
+                break;
+            }
+
+            // Sleep for some time before trying again
+            await new Promise((resolve) => setTimeout(resolve, this.httpClient.minDelayBetweenRetriesMillis));
+        }
+
+        // TODO: Not sure how to do this properly in ts-friendly way
+        const result = <unknown> { processedRequests, unprocessedRequests } as JsonValue;
+
+        return cast(parseDateFields(result));
+    }
+
+    async batchAddRequests(
+        requests: Array<Omit<RequestQueueClientRequestSchema, 'id'>>,
+        options: RequestQueueClientBatchAddRequestOptions = {},
+    ): Promise<RequestQueueClientBatchAddRequestsResult> {
+        ow(requests, ow.array.ofType(ow.object.partialShape({
+            id: ow.undefined,
+        })).minLength(1));
+
+        ow(options, ow.object.exactShape({
+            forefront: ow.optional.boolean,
+        }));
+
+        // The number of 50 parallel requests seemed optimal, if it was higher it did not seem to bring any extra value.
+        const maxParallelRequests = 50;
+
+        const operationsInProgress = [];
+        const individualResults = [];
+
+        // Send up to `maxParallelRequests` requests at once, wait for all of them to finish and repeat
+        for (let i = 0; i < requests.length; i += 25) {
+            const requestsInBatch = requests.slice(i, i + 25);
+            operationsInProgress.push(this._batchAddRequestsWithRetries(requestsInBatch, options));
+            if (operationsInProgress.length === maxParallelRequests) {
+                individualResults.push(...(await Promise.all(operationsInProgress)));
+                operationsInProgress.splice(0, operationsInProgress.length);
```

Comment:
This seems that if any of the batches would throw, the whole thing would throw. But in background, more batches will still be running with unknown states.

Reply:
Did not think of this. Maybe it could be dealt with in `_batchAddRequestsWithRetries`, make it never throwing and just take request processed so far into the account.